### ISSUE TITLE
divergence in protocol variance inference between ty and pyright

### ISSUE THREAD

**Issue by carljm** [astral-sh/ty#4407]
> an underlying issue that may lead to a lot of divergence between my "baseline" basedpyright and ty is that the former accept `PyoIterator` and `Iter` as covariant whilst not ty nor pyrefly do. I'm not sure who's right tho.
> I tried smaller reproducible examples but they were working as expected (i.e all 3 checkers converging).
> 
> Examples are below. Do you want me to raise another issue? 
> 
> ---
> 
> ## First example
> 
> ```python
> from __future__ import annotations
> 
> from typing import TYPE_CHECKING, assert_type
> 
> from pyochain import Iter
> 
> if TYPE_CHECKING:
>     from collections.abc import Iterator
> 
>     from pyochain.abc import PyoIterator
> 
> 
> class Animal: ...
> 
> 
> class Dog(Animal): ...
> 
> 
> a = assert_type(Iter(Dog()), Iter[Dog])
> _b: Iter[Animal] = a
> _c: PyoIterator[Animal] = a
> _d: Iterator[Animal] = a
> ```
> ### ty output
> ```python
> PS C:\Users\stett\Documents\python\pyochain> uvx ty check t.py
> error[invalid-assignment]: Object of type `Iter[Dog]` is not assignable to `Iter[Animal]`
>   --> t.py:20:20
>    |
> 20 | _b: Iter[Animal] = a
>    |     ------------   ^ Incompatible value of type `Iter[Dog]`
>    |     |
>    |     Declared type
> info: `Iter` is invariant in its type parameter
> info: For more information, see https://docs.astral.sh/ty/reference/typing-faq/#invariant-generics
> 
> error[invalid-assignment]: Object of type `Iter[Dog]` is not assignable to `PyoIterator[Animal]`
>   --> t.py:21:27
>    |
> 21 | _c: PyoIterator[Animal] = a
>    |     -------------------   ^ Incompatible value of type `Iter[Dog]`
>    |     |
>    |     Declared type
> info: type `Iter[Dog]` is not assignable to protocol `PyoIterator[Animal]`
> info: └── protocol member `batched` is incompatible
> 
> Found 2 diagnostics
> ```
> ### Pyrefly output
> ```shell
> PS C:\Users\stett\Documents\python\pyochain> uvx pyrefly check t.py  
> ERROR `Iter[Dog]` is not assignable to `Iter[Animal]` [bad-assignment]
>   --> t.py:20:20
>    |
> 20 | _b: Iter[Animal] = a
>    |     ------------   ^
>    |     |
>    |     declared type
>    |
> ERROR `Iter[Dog]` is not assignable to `PyoIterator[Animal]` [bad-assignment]
>   --> t.py:21:27
>    |
> 21 | _c: PyoIterator[Animal] = a
>    |     -------------------   ^
>    |     |
>    |     declared type
>    |
>  INFO 2 errors
> ```
> ### Basedpyright output
> ```shell
> PS C:\Users\stett\Documents\python\pyochain> uv run basedpyright t.py
> 0 errors, 0 warnings, 0 notes
> PS C:\Users\stett\Documents\python\pyochain> 
> ```
> ## Second example
> 
> Replacing the iterators by sequences (may be easier to pinpoint where it comes from, since the number of methods is much smaller)
> ```python
> # Swap from pyochain import Iter to from pyochain import Seq
> a = assert_type(Seq(Dog()), Seq[Dog])
> _b: Seq[Animal] = a
> _c: PyoSequence[Animal] = a
> _d: Sequence[Animal] = a
> ```
> ### ty output
> ```shell
> PS C:\Users\stett\Documents\python\pyochain> uvx ty check t.py       
> error[invalid-assignment]: Object of type `Seq[Dog]` is not assignable to `Seq[Animal]`
>   --> t.py:20:19
>    |
> 20 | _b: Seq[Animal] = a
>    |     -----------   ^ Incompatible value of type `Seq[Dog]`
>    |     |
>    |     Declared type
> info: `Seq` is invariant in its type parameter
> info: For more information, see https://docs.astral.sh/ty/reference/typing-faq/#invariant-generics
> 
> error[invalid-assignment]: Object of type `Seq[Dog]` is not assignable to `PyoSequence[Animal]`
>   --> t.py:21:27
>    |
> 21 | _c: PyoSequence[Animal] = a
>    |     -------------------   ^ Incompatible value of type `Seq[Dog]`
>    |     |
>    |     Declared type
> 
> Found 2 diagnostics
> ```
> ### pyrefly output
> ```shell
> PS C:\Users\stett\Documents\python\pyochain> uvx pyrefly check t.py
> ERROR `Seq[Dog]` is not assignable to `Seq[Animal]` [bad-assignment]
>   --> t.py:20:19
>    |
> 20 | _b: Seq[Animal] = a
>    |     -----------   ^
>    |     |
>    |     declared type
>    |
> ERROR `Seq[Dog]` is not assignable to `PyoSequence[Animal]` [bad-assignment]
>   --> t.py:21:27
>    |
> 21 | _c: PyoSequence[Animal] = a
>    |     -------------------   ^
>    |     |
>    |     declared type
>    |
>  INFO 2 errors
> ```
> ### basedpyright output
> ```shell
> PS C:\Users\stett\Documents\python\pyochain> uv run basedpyright t.py
> 0 errors, 0 warnings, 0 notes
> PS C:\Users\stett\Documents\python\pyochain> 
> ``` 

 _Originally posted by @OutSquareCapital in [#4269](https://github.com/astral-sh/ty/issues/4269#issuecomment-5443898863)_

**Comment by carljm** [comment]
Thanks for the report! There are two contributing limitations here:

1. Recursive protocols currently fall back to a form of variance inference that wrongly exposes raw signatures with `self` type still present. `ResultType` has its class type parameter `E` in several receiver annotations, wrongly contributing what looks like a contravariant use. https://github.com/astral-sh/ruff/pull/28077 fixes this issue.
2. We currently lose `Final` qualifiers during variance inference in stubs. We wrongly consider even this simple `Box` type to be invariant, if defined in a stub:

```py
# box.pyi
from typing import Final

class Box[T]:
    value: Final[T]
```

**Comment by carljm** [comment]
The Final-in-stubs issue is now fixed on main, leaving only the recursive-protocol part of this issue, which will be fixed by https://github.com/astral-sh/ruff/pull/28077

**Comment by OutSquareCapital** [comment]
Fantastic! ty issue resolving speed seems on par with the type checking speed

**Comment by OutSquareCapital** [comment]
Has this landed in 0.0.78, or yet to be released? I unfortunately still see errors in my example

**Comment by charliermarsh** [comment]
It's yet to be released.
